File: docs/palindromes-in-peg/scaffold_circuit.py
"""Predicated finite-value and local-pointer equations for scaffold lowering.

This is a construction-time API. Its Python dictionaries are finite expression
maps; the emitted machine has only Boolean labels, pointer fields, and ordinary
PEG rules. Ref distinguishes a cell allocated during this tick from an old cell
so reading a new cell uses its construction-time fields, never recursive queries
of the new scaffold node. No pointer identity operation is supplied.
"""
from dataclasses import dataclass

from symbolic_sca2peg import (Scaffold, Expr, TRUE, FALSE, SELF, NULL, old, symbol,
                              pointer, read, edge, present, select)


def neg(a):
  if a == TRUE: return FALSE
  if a == FALSE: return TRUE
  if a[0] == "not": return a[1]
  return Expr(("not", a))


def conjunction(*args):
  result, seen = [], set()
  for arg in args:
    if arg == FALSE: return FALSE
    if arg != TRUE and arg not in seen:
      result.append(arg); seen.add(arg)
  return FALSE if any(neg(x) in seen for x in result) else \
    TRUE if not result else result[0] if len(result) == 1 else Expr(("and", *result))


def disjunction(*args):
  result, seen = [], set()
  for arg in args:
    if arg == TRUE: return TRUE
    if arg != FALSE and arg not in seen:
      result.append(arg); seen.add(arg)
  return TRUE if any(neg(x) in seen for x in result) else \
    FALSE if not result else result[0] if len(result) == 1 else Expr(("or", *result))


def choose(guard, yes, no):
  if guard == TRUE or yes == no: return yes
  if guard == FALSE: return no
  if yes == TRUE and no == FALSE: return guard
  if yes == FALSE and no == TRUE: return neg(guard)
  return disjunction(conjunction(guard, yes), conjunction(neg(guard), no))


def choose_pointer(guard, yes, no):
  if guard == TRUE or yes == no: return yes
  if guard == FALSE: return no
  return select(guard, yes, no)


class Value:
  """A finite tagged value with a shared binary encoding and a validity bit."""
  def __init__(self, cases):
    cases = {value: guard for value, guard in cases.items() if guard != FALSE}
    self.domain = tuple(cases)
    self.valid = disjunction(*cases.values())
    self.bits = tuple(disjunction(*(guard for index, guard in enumerate(cases.values())
                                   if index & (1 << bit)))
                      for bit in range(max(0, (len(cases) - 1).bit_length())))

  @classmethod
  def encoded(cls, domain, bits, valid=TRUE):
    result = object.__new__(cls)
    result.domain, result.bits, result.valid = tuple(domain), tuple(bits), valid
    return result

  @classmethod
  def constant(cls, value):
    return cls({value: TRUE})

  @property
  def cases(self):
    return {value: self.eq(value) for value in self.domain}

  def eq(self, value):
    if value not in self.domain: return FALSE
    index = self.domain.index(value)
    return conjunction(self.valid, *(bit if index & (1 << i) else neg(bit)
                                     for i, bit in enumerate(self.bits)))

  def recode(self, domain):
    domain = tuple(domain)
    if domain == self.domain: return self
    if not set(self.domain) <= set(domain):
      raise ValueError("cannot discard variants of a finite value")
    bits = tuple(disjunction(*(self.eq(value) for index, value in enumerate(domain)
                               if index & (1 << bit)))
                 for bit in range(max(0, (len(domain) - 1).bit_length())))
    return Value.encoded(domain, bits, self.valid)

  def map(self, function):
    result = {}
    for value, guard in self.cases.items():
      mapped = function(value)
      result[mapped] = disjunction(result.get(mapped, FALSE), guard)
    return Value(result)

  def cycle(self, direction=1):
    """Rotate a power-of-two domain by one with a binary carry circuit."""
    if direction not in (-1, 1) or not self.domain or len(self.domain) & (len(self.domain) - 1):
      raise ValueError("unit cyclic movement requires a power-of-two domain")
    carry, bits = TRUE, []
    for bit in self.bits:
      bits.append(choose(carry, neg(bit), bit))
      carry = conjunction(carry, bit if direction == 1 else neg(bit))
    return Value.encoded(self.domain, bits, self.valid)

  def equal(self, other):
    if self.domain == other.domain:
      return conjunction(self.valid, other.valid,
        *(choose(a, b, neg(b)) for a, b in zip(self.bits, other.bits)))
    return disjunction(*(conjunction(guard, other.eq(value))
                         for value, guard in self.cases.items()))

  @staticmethod
  def select(guard, yes, no):
    if guard == TRUE: return yes
    if guard == FALSE: return no
    domain = tuple(dict.fromkeys((*no.domain, *yes.domain)))
    yes, no = yes.recode(domain), no.recode(domain)
    return Value.encoded(domain, (choose(guard, a, b) for a, b in zip(yes.bits, no.bits)),
                         choose(guard, yes.valid, no.valid))


@dataclass(frozen=True)
class Ref:
  new: tuple = FALSE
  prior: tuple = NULL

  def present(self):
    return disjunction(self.new, present(self.prior) if self.prior != NULL else FALSE)

  @staticmethod
  def select(guard, yes, no):
    return Ref(choose(guard, yes.new, no.new),
               choose_pointer(guard, yes.prior, no.prior))

  def expression(self):
    return choose_pointer(self.new, SELF, self.prior)


NEW = Ref(TRUE, NULL)
EMPTY = Ref()
PREVIOUS = Ref(FALSE, pointer(()))


class Circuit:
  def __init__(self, alphabet="ab", check_invariants=True):
    self.alphabet = alphabet
    self.check_invariants = check_invariants
    self.domains, self.initial, self.labels, self.pointers = {}, {}, {}, {}
    self.current_values, self.current_refs = {}, {}
    self.label_ids = {}

  def _label(self, key, value):
    pair = key, value
    if pair not in self.label_ids:
      self.label_ids[pair] = f"v{len(self.label_ids)}"
    return self.label_ids[pair]

  def scalar(self, key, domain, initial):
    domain = tuple(domain)
    if initial not in domain or len(set(domain)) != len(domain):
      raise ValueError("finite scalar domain must be distinct and include its initial value")
    if key in self.domains:
      if self.domains[key] != (domain, initial):
        raise ValueError(f"inconsistent scalar declaration: {key}")
      return
    self.domains[key] = domain, initial
    for bit in range(max(0, (len(domain) - 1).bit_length())):
      name = self._label(key, bit)
      self.initial[name] = bool(domain.index(initial) & (1 << bit))
      self.labels[name] = FALSE

  def get(self, target, key, domain, initial):
    self.scalar(key, domain, initial)
    current = self.current_values.get(key, Value.constant(initial)).recode(domain)
    bits = tuple(choose(target.new, current.bits[bit],
                        read(target.prior, self._label(key, bit)))
                 for bit in range(len(current.bits)))
    return Value.encoded(domain, bits, target.present())

  def put(self, key, value, domain=None, initial=None):
    if key not in self.domains:
      if domain is None: raise ValueError(f"undeclared scalar: {key}")
      self.scalar(key, domain, initial)
    value = value.recode(self.domains[key][0])
    self.current_values[key] = value
    for bit, expr in enumerate(value.bits):
      self.labels[self._label(key, bit)] = expr

  def get_ref(self, target, key):
    self.pointers.setdefault(key, NULL)
    local = self.current_refs.get(key, EMPTY)
    return Ref(choose(target.new, local.new, FALSE),
               choose_pointer(target.new, local.prior, edge(target.prior, key)))

  def put_ref(self, key, value):
    self.current_refs[key] = value
    self.pointers[key] = value.expression()

  def input(self):
    return Value({char: symbol(char) for char in self.alphabet})

  def require(self, condition, enabled=TRUE):
    if not self.check_invariants: return
    if not hasattr(self, "fault"):
      self.fault = self.get(PREVIOUS, "circuit.fault", (False, True), False).eq(True)
    self.fault = disjunction(self.fault, conjunction(enabled, neg(condition)))

  def machine(self, accepting, initial_accepting=False):
    if hasattr(self, "fault"):
      self.put("circuit.fault", Value.select(self.fault, Value.constant(True), Value.constant(False)))
      accepting = conjunction(accepting, neg(self.fault))
    # Unwritten cell fields have their declared defaults; this matters when a
    # later node aliases a slot never allocated in a particular branch.
    for key, (domain, initial) in self.domains.items():
      if key not in self.current_values:
        for bit in range(max(0, (len(domain) - 1).bit_length())):
          self.labels[self._label(key, bit)] = TRUE if domain.index(initial) & (1 << bit) else FALSE
    name = "accept"
    if name in self.initial: raise ValueError("reserved acceptance label")
    self.initial[name] = initial_accepting
    self.labels[name] = accepting
    return Scaffold(self.initial, self.labels, self.pointers, name, self.alphabet)

File: scala/pal/src/main/scala/pal/ScaffoldCircuit.scala
package pal

import scala.collection.mutable

import Expr.{TRUE, FALSE, SELF, NULL}

/** Predicated finite-value and local-pointer equations for scaffold lowering.
  *
  * This is a construction-time API. Its finite expression maps are consumed at
  * construction; the emitted machine has only Boolean labels, pointer fields,
  * and ordinary PEG rules. `Ref` distinguishes a cell allocated during this
  * tick from an old cell so reading a new cell uses its construction-time
  * fields, never recursive queries of the new scaffold node. No pointer
  * identity operation is supplied.
  *
  * Port of `scaffold_circuit.py`. The Boolean simplifiers (`neg`, `conjunction`,
  * `disjunction`, `choose`, `choosePointer`) live here; `Value`, `Ref` and
  * `Circuit` are top-level classes of package `pal`.
  */
object ScaffoldCircuit {

  def neg(a: Expr): Expr = {
    a match {
      case Expr.Const(true) => FALSE
      case Expr.Const(false) => TRUE
      case Expr.Not(inner) => inner
      case _ => Expr.intern(Expr.Not(a))
    }
  }

  /** The distinct non-`unit` operands in first-occurrence order plus their set, or `None` when `zero` short-circuits. */
  private def operands(args: Seq[Expr], zero: Expr, unit: Expr): Option[(Vector[Expr], mutable.HashSet[Expr])] = {
    val result = Vector.newBuilder[Expr]
    val seen = mutable.HashSet.empty[Expr]
    var shortCircuit = false
    val remaining = args.iterator
    while (!shortCircuit && remaining.hasNext) {
      val arg = remaining.next()
      if (arg == zero) {
        shortCircuit = true
      } else if (arg != unit && seen.add(arg)) {
        result += arg
      }
    }
    if (shortCircuit) { None } else { Some((result.result(), seen)) }
  }

  /** Shared body of `conjunction` (`isOr = false`) and `disjunction` (`isOr = true`). */
  private def connective(args: Seq[Expr], isOr: Boolean): Expr = {
    val zero = if (isOr) { TRUE } else { FALSE }
    val unit = if (isOr) { FALSE } else { TRUE }
    operands(args, zero, unit) match {
      case None => zero
      case Some((result, seen)) =>
        if (result.exists(x => seen.contains(neg(x)))) {
          zero
        } else if (result.isEmpty) {
          unit
        } else if (result.length == 1) {
          result.head
        } else if (isOr) {
          Expr.intern(Expr.Or(result))
        } else {
          Expr.intern(Expr.And(result))
        }
    }
  }

  def conjunction(args: Expr*): Expr = connective(args, isOr = false)

  def disjunction(args: Expr*): Expr = connective(args, isOr = true)

  def choose(guard: Expr, yes: Expr, no: Expr): Expr = {
    if (guard == TRUE || yes == no) {
      yes
    } else if (guard == FALSE) {
      no
    } else if (yes == TRUE && no == FALSE) {
      guard
    } else if (yes == FALSE && no == TRUE) {
      neg(guard)
    } else {
      disjunction(conjunction(guard, yes), conjunction(neg(guard), no))
    }
  }

  def choosePointer(guard: Expr, yes: Expr, no: Expr): Expr = {
    if (guard == TRUE || yes == no) {
      yes
    } else if (guard == FALSE) {
      no
    } else {
      Expr.select(guard, yes, no)
    }
  }

  /** Python `max(0, (size - 1).bit_length())`: bits needed to index `size` variants.
    *
    * Note `(-1).bit_length() == 1`, so an empty domain still gets one (false) bit.
    */
  private[pal] def bitWidth(size: Int): Int = {
    if (size <= 0) { 1 } else { 32 - Integer.numberOfLeadingZeros(size - 1) }
  }
}

/** A finite tagged value with a shared binary encoding and a validity bit.
  *
  * Domain elements are arbitrary (Python used characters, Booleans, integers,
  * strings, `None` and tag tuples); Scala `None` plays the role of Python `None`.
  */
final class Value[+A] private (val domain: Vector[A], val bits: Vector[Expr], val valid: Expr) {
  import ScaffoldCircuit.*

  /** Position of each domain element (Python `domain.index`, but O(1) after the first use). */
  private lazy val positions: Map[Any, Int] = domain.zipWithIndex.toMap[Any, Int]

  /** `(value, guard)` pairs in domain order (Python `cases` dict). */
  def cases: Vector[(A, Expr)] = domain.map(value => value -> eqTo(value))

  /** Guard for `this == value` (Python `Value.eq`; renamed because `AnyRef.eq` is final). */
  def eqTo(value: Any): Expr = {
    positions.get(value) match {
      case None => FALSE
      case Some(index) =>
        val tests = bits.zipWithIndex.map { case (bit, i) => if ((index & (1 << i)) != 0) { bit } else { neg(bit) } }
        conjunction((valid +: tests)*)
    }
  }

  def recode[B >: A](target: Seq[B]): Value[B] = {
    val domain = target.toVector
    if (domain == this.domain) {
      this
    } else {
      val available = domain.toSet
      if (!this.domain.forall(value => available.contains(value))) {
        throw new IllegalArgumentException("cannot discard variants of a finite value")
      }
      val bits = (0 until bitWidth(domain.length)).map { bit =>
        disjunction(domain.zipWithIndex.collect { case (value, index) if (index & (1 << bit)) != 0 => eqTo(value) }*)
      }
      Value.encoded(domain, bits, valid)
    }
  }

  def map[B](function: A => B): Value[B] = {
    // A LinkedHashMap keeps the position of the first occurrence when a key is
    // merged again, exactly like the Python dict the original accumulated into.
    val result = mutable.LinkedHashMap.empty[B, Expr]
    for ((value, guard) <- cases) {
      val mapped = function(value)
      result(mapped) = disjunction(result.getOrElse(mapped, FALSE), guard)
    }
    Value(result)
  }

  /** Rotate a power-of-two domain by one with a binary carry circuit. */
  def cycle(direction: Int = 1): Value[A] = {
    val size = domain.length
    if ((direction != -1 && direction != 1) || size == 0 || (size & (size - 1)) != 0) {
      throw new IllegalArgumentException("unit cyclic movement requires a power-of-two domain")
    }
    var carry = TRUE
    val rotated = Vector.newBuilder[Expr]
    for (bit <- bits) {
      rotated += choose(carry, neg(bit), bit)
      carry = conjunction(carry, if (direction == 1) { bit } else { neg(bit) })
    }
    Value.encoded(domain, rotated.result(), valid)
  }

  def equal(other: Value[?]): Expr = {
    if (domain == other.domain) {
      val agree = bits.zip(other.bits).map { case (a, b) => choose(a, b, neg(b)) }
      conjunction((Vector(valid, other.valid) ++ agree)*)
    } else {
      disjunction(cases.map { case (value, guard) => conjunction(guard, other.eqTo(value)) }*)
    }
  }

  override def toString: String = s"Value(domain=$domain, bits=$bits, valid=$valid)"
}

object Value {
  import ScaffoldCircuit.*

  /** Build from `(value, guard)` cases in order; `FALSE` guards are dropped. */
  def apply[A](cases: Iterable[(A, Expr)]): Value[A] = {
    val merged = mutable.LinkedHashMap.empty[A, Expr]
    for ((value, guard) <- cases) { merged(value) = guard }
    val live = merged.toVector.filter { case (_, guard) => guard != FALSE }
    val domain = live.map(_._1)
    val valid = disjunction(live.map(_._2)*)
    val bits = (0 until bitWidth(live.length)).map { bit =>
      disjunction(live.zipWithIndex.collect { case ((_, guard), index) if (index & (1 << bit)) != 0 => guard }*)
    }
    encoded(domain, bits, valid)
  }

  def encoded[A](domain: Seq[A], bits: Iterable[Expr], valid: Expr = TRUE): Value[A] = {
    new Value(domain.toVector, bits.toVector, valid)
  }

  def constant[A](value: A): Value[A] = apply(Vector(value -> TRUE))

  def select[A](guard: Expr, yes: Value[A], no: Value[A]): Value[A] = {
    if (guard == TRUE) {
      yes
    } else if (guard == FALSE) {
      no
    } else {
      val domain = (no.domain ++ yes.domain).distinct
      val y = yes.recode(domain)
      val n = no.recode(domain)
      encoded(domain, y.bits.zip(n.bits).map { case (a, b) => choose(guard, a, b) }, choose(guard, y.valid, n.valid))
    }
  }
}

/** A cell reference: `isNew` when allocated this tick, else the old pointer `prior`. */
final case class Ref(isNew: Expr = FALSE, prior: Expr = NULL) {
  import ScaffoldCircuit.*

  def present(): Expr = disjunction(isNew, if (prior != NULL) { Expr.present(prior) } else { FALSE })

  def expression(): Expr = choosePointer(isNew, SELF, prior)
}

object Ref {
  import ScaffoldCircuit.*

  def select(guard: Expr, yes: Ref, no: Ref): Ref = {
    Ref(choose(guard, yes.isNew, no.isNew), choosePointer(guard, yes.prior, no.prior))
  }

  val NEW: Ref = Ref(TRUE, NULL)
  val EMPTY: Ref = Ref()
  val PREVIOUS: Ref = Ref(FALSE, Expr.pointer(Nil))
}

/** Accumulates finite scalar and pointer equations, then emits a `Scaffold`. */
final class Circuit(val alphabet: String = "ab", val checkInvariants: Boolean = true) {
  import ScaffoldCircuit.*

  /** Declared scalar domains: key -> (domain, initial value). */
  val domains: mutable.LinkedHashMap[String, (Vector[Any], Any)] = mutable.LinkedHashMap.empty
  val initial: mutable.LinkedHashMap[String, Boolean] = mutable.LinkedHashMap.empty
  val labels: mutable.LinkedHashMap[String, Expr] = mutable.LinkedHashMap.empty
  val pointers: mutable.LinkedHashMap[String, Expr] = mutable.LinkedHashMap.empty
  val currentValues: mutable.LinkedHashMap[String, Value[Any]] = mutable.LinkedHashMap.empty
  val currentRefs: mutable.LinkedHashMap[String, Ref] = mutable.LinkedHashMap.empty
  val labelIds: mutable.LinkedHashMap[(String, Int), String] = mutable.LinkedHashMap.empty

  /** Accumulated invariant violations (Python's lazily created `fault` attribute). */
  var fault: Option[Expr] = None

  /** Cell pools registered by `StackPool` (Python's dynamically attached `_pools`). */
  val pools: mutable.ArrayBuffer[StackPool] = mutable.ArrayBuffer.empty

  private def label(key: String, bit: Int): String = labelIds.getOrElseUpdate((key, bit), s"v${labelIds.size}")

  def scalar[A](key: String, domain: Seq[A], initial: A): Unit = {
    val values: Vector[Any] = domain.toVector
    if (!values.contains(initial) || values.distinct.length != values.length) {
      throw new IllegalArgumentException("finite scalar domain must be distinct and include its initial value")
    }
    domains.get(key) match {
      case Some(declared) =>
        if (declared != ((values, initial))) { throw new IllegalArgumentException(s"inconsistent scalar declaration: $key") }
      case None =>
        domains(key) = (values, initial)
        val index = values.indexOf(initial)
        for (bit <- 0 until bitWidth(values.length)) {
          val name = label(key, bit)
          this.initial(name) = (index & (1 << bit)) != 0
          labels(name) = FALSE
        }
    }
  }

  def get[A](target: Ref, key: String, domain: Seq[A], initial: A): Value[A] = {
    scalar(key, domain, initial)
    val current = currentValues.getOrElse(key, Value.constant(initial)).recode(domain)
    val bits = current.bits.zipWithIndex.map { case (bit, i) => choose(target.isNew, bit, Expr.read(target.prior, label(key, i))) }
    Value.encoded(domain, bits, target.present())
  }

  def put(key: String, value: Value[Any]): Unit = {
    if (!domains.contains(key)) { throw new IllegalArgumentException(s"undeclared scalar: $key") }
    store(key, value)
  }

  def put[A](key: String, value: Value[A], domain: Seq[A], initial: A): Unit = {
    if (!domains.contains(key)) { scalar(key, domain, initial) }
    store(key, value)
  }

  private def store(key: String, value: Value[Any]): Unit = {
    val recoded = value.recode(domains(key)._1)
    currentValues(key) = recoded
    for ((expr, bit) <- recoded.bits.zipWithIndex) { labels(label(key, bit)) = expr }
  }

  def getRef(target: Ref, key: String): Ref = {
    pointers.getOrElseUpdate(key, NULL)
    val local = currentRefs.getOrElse(key, Ref.EMPTY)
    Ref(choose(target.isNew, local.isNew, FALSE), choosePointer(target.isNew, local.prior, Expr.edge(target.prior, key)))
  }

  def putRef(key: String, value: Ref): Unit = {
    currentRefs(key) = value
    pointers(key) = value.expression()
  }

  def input(): Value[Char] = Value(alphabet.toVector.map(char => char -> Expr.symbol(char)))

  def require(condition: Expr, enabled: Expr = TRUE): Unit = {
    if (checkInvariants) {
      val current = fault.getOrElse(get(Ref.PREVIOUS, "circuit.fault", Vector(false, true), false).eqTo(true))
      fault = Some(disjunction(current, conjunction(enabled, neg(condition))))
    }
  }

  def machine(accepting: Expr, initialAccepting: Boolean = false): Scaffold = {
    val accept = fault match {
      case Some(current) =>
        put("circuit.fault", Value.select(current, Value.constant(true), Value.constant(false)))
        conjunction(accepting, neg(current))
      case None => accepting
    }
    // Unwritten cell fields have their declared defaults; this matters when a
    // later node aliases a slot never allocated in a particular branch.
    for ((key, (domain, initial)) <- domains if !currentValues.contains(key)) {
      val index = domain.indexOf(initial)
      for (bit <- 0 until bitWidth(domain.length)) {
        labels(label(key, bit)) = if ((index & (1 << bit)) != 0) { TRUE } else { FALSE }
      }
    }
    val name = "accept"
    if (initial.contains(name)) { throw new IllegalArgumentException("reserved acceptance label") }
    initial(name) = initialAccepting
    labels(name) = accept
    new Scaffold(initial, labels, pointers, name, alphabet)
  }
}
